File: apps/web/src/lib/ai/summarizer.ts
/**
 * Message Summarizer Module
 *
 * Standalone functions for summarizing conversations,
 * extracting key points, action items, and decisions.
 *
 * @module lib/ai/summarizer
 */

import type { MessageSummary } from './types';

/**
 * Score and extract key points from sentences
 */
export function extractKeyPoints(sentences: string[]): string[] {
  // Score sentences by importance
  const scored = sentences.map((sentence) => {
    let score = 0;

    // Longer sentences often contain more info
    score += Math.min(sentence.split(' ').length / 20, 1) * 0.3;

    // Sentences with numbers are often important
    if (/\d+/.test(sentence)) score += 0.2;

    // Sentences with quotes
    if (/["']/.test(sentence)) score += 0.15;

    // Sentences with emphasis
    if (/!/.test(sentence)) score += 0.1;

    // Key phrases
    if (/\b(important|critical|key|main|significant|essential)\b/i.test(sentence)) {
      score += 0.3;
    }

    return { sentence: sentence.trim(), score };
  });

  return scored
    .sort((a, b) => b.score - a.score)
    .slice(0, 5)
    .map((s) => s.sentence);
}

/**
 * Generate a one-line brief summary
 */
export function generateBriefSummary(messages: Array<{ sender: string; content: string }>): string {
  const participants = [...new Set(messages.map((m) => m.sender))];
  const messageCount = messages.length;

  if (messageCount === 1) {
    return `${participants[0]} sent a message.`;
  }

  return `Conversation between ${participants.join(' and ')} with ${messageCount} messages.`;
}

/**
 * Generate a detailed paragraph summary
 */
export function generateDetailedSummary(
  messages: Array<{ sender: string; content: string }>,
  keyPoints: string[]
): string {
  const participants = [...new Set(messages.map((m) => m.sender))];

  let summary = `This conversation involves ${participants.join(', ')}. `;

  if (keyPoints.length > 0) {
    summary += `Key topics discussed include: ${keyPoints.slice(0, 3).join('; ')}. `;
  }

  summary += `Total of ${messages.length} messages exchanged.`;

  return summary;
}

/**
 * Summarize a conversation into structured output
 */
export async function summarizeConversation(
  messages: Array<{ sender: string; content: string; timestamp: number }>
): Promise<MessageSummary> {
  if (messages.length === 0) {
    return {
      brief: 'No messages to summarize.',
      detailed: 'The conversation is empty.',
      keyPoints: [],
      actionItems: [],
      decisions: [],
      questions: [],
    };
  }

  // Extract key information
  const allText = messages.map((m) => m.content).join(' ');
  const sentences = allText.split(/[.!?]+/).filter((s) => s.trim().length > 0);

  // Extract action items (sentences with action verbs)
  const actionItems = sentences
    .filter((s) => /\b(need to|should|must|will|going to|have to|let'?s)\b/i.test(s))
    .map((s) => s.trim())
    .slice(0, 5);

  // Extract questions
  const questions = messages
    .filter((m) => m.content.includes('?'))
    .map((m) => m.content)
    .slice(0, 5);

  // Extract decisions (sentences with decision indicators)
  const decisions = sentences
    .filter((s) => /\b(decided|agreed|confirmed|approved|will do|let'?s go with)\b/i.test(s))
    .map((s) => s.trim())
    .slice(0, 5);

  // Generate key points (most important sentences)
  const keyPoints = extractKeyPoints(sentences);

  // Generate summaries
  const brief = generateBriefSummary(messages);
  const detailed = generateDetailedSummary(messages, keyPoints);

  return {
    brief,
    detailed,
    keyPoints,
    actionItems,
    decisions,
    questions,
  };
}

File: apps/backend/lib/cgraph/ai/summarizer.ex
defmodule CGraph.AI.Summarizer do
  @moduledoc """
  Conversation summarization — LLM-powered with heuristic fallback.
  """

  alias CGraph.AI.LLMClient

  @doc """
  Summarize a conversation into structured output.

  Returns:
    %{
      brief: "One-line summary",
      detailed: "Multi-paragraph summary",
      key_points: ["point 1", ...],
      action_items: ["action 1", ...],
      decisions: ["decision 1", ...]
    }
  """
  def summarize(messages, opts \\ []) when is_list(messages) do
    if CGraph.AI.llm_available?() do
      llm_summarize(messages, opts)
    else
      heuristic_summarize(messages)
    end
  end

  # ---------------------------------------------------------------------------
  # LLM Implementation
  # ---------------------------------------------------------------------------

  defp llm_summarize(messages, opts) do
    system_prompt = """
    You are a conversation summarizer. Given a conversation, produce a JSON response with:
    - "brief": A one-sentence summary
    - "detailed": A 2-3 paragraph detailed summary
    - "key_points": Array of 3-5 key points
    - "action_items": Array of action items mentioned
    - "decisions": Array of decisions made

    Respond ONLY with valid JSON.
    """

    formatted =
      messages
      |> Enum.map(fn msg ->
        sender = msg["sender"] || msg[:sender] || "Unknown"
        content = msg["content"] || msg[:content] || ""
        "#{sender}: #{content}"
      end)
      |> Enum.join("\n")

    llm_messages = [
      %{role: "system", content: system_prompt},
      %{role: "user", content: "Summarize this conversation:\n\n#{formatted}"}
    ]

    case LLMClient.complete(llm_messages, opts) do
      {:ok, response} ->
        case Jason.decode(response) do
          {:ok, parsed} ->
            {:ok, %{
              brief: parsed["brief"] || "",
              detailed: parsed["detailed"] || "",
              key_points: parsed["key_points"] || [],
              action_items: parsed["action_items"] || [],
              decisions: parsed["decisions"] || []
            }}

          {:error, _} ->
            # If JSON parsing fails, wrap raw response
            {:ok, %{
              brief: String.slice(response, 0, 200),
              detailed: response,
              key_points: [],
              action_items: [],
              decisions: []
            }}
        end

      {:error, _reason} ->
        # Fallback to heuristic
        heuristic_summarize(messages)
    end
  end

  # ---------------------------------------------------------------------------
  # Heuristic Fallback
  # ---------------------------------------------------------------------------

  defp heuristic_summarize(messages) do
    participants =
      messages
      |> Enum.map(&(Map.get(&1, "sender") || Map.get(&1, :sender, "Unknown")))
      |> Enum.uniq()

    all_text =
      messages
      |> Enum.map(&(Map.get(&1, "content") || Map.get(&1, :content, "")))
      |> Enum.join(" ")

    sentences =
      all_text
      |> String.split(~r/[.!?]+/)
      |> Enum.map(&String.trim/1)
      |> Enum.reject(&(&1 == ""))

    action_items =
      sentences
      |> Enum.filter(&Regex.match?(~r/\b(need to|should|must|will|going to|have to|let's)\b/i, &1))
      |> Enum.take(5)

    decisions =
      sentences
      |> Enum.filter(&Regex.match?(~r/\b(decided|agreed|confirmed|approved|will do|let's go with)\b/i, &1))
      |> Enum.take(5)

    key_points = Enum.take(sentences, 5)

    brief = "Conversation between #{Enum.join(participants, " and ")} with #{length(messages)} messages."

    detailed =
      "This conversation involves #{Enum.join(participants, ", ")}. " <>
      if(length(key_points) > 0,
        do: "Key topics: #{Enum.take(key_points, 3) |> Enum.join("; ")}. ",
        else: "") <>
      "Total of #{length(messages)} messages exchanged."

    {:ok, %{
      brief: brief,
      detailed: detailed,
      key_points: key_points,
      action_items: action_items,
      decisions: decisions
    }}
  end
end
